torchx/c_src/torchx.cpp

```diff
@@ -136,10 +141,17 @@ NIF(delete_tensor)
 {
   TENSOR_PARAM(0, t);
 
-  t->~Tensor();
-  enif_release_resource(t);
+  if (*(intptr_t*)t != EMPTY_TAG)
+  {
+    t->~Tensor();
+    *(intptr_t*)t = EMPTY_TAG;
 
-  return nx::nif::ok(env);
+    return nx::nif::ok(env);
+  }
+  else
+  {
+    return nx::nif::error(env, "already deleted tensor");
```

Comment:
Do we ever get to this clause? 🤔

Reply:
we should not as we already checked in TENSOR_PARAM macro. But I don't know what is proper way to express that here.

Reply:
One option would be to remove the conditional altogether and keep only the "if" branch. If you don't want to do that, then I would add a comment:

```suggestion
    // We should never reach this clause as TENSOR_PARAM would have raised
    return nx::nif::error(env, "already deleted tensor");
```

I have also just realized that this code can have race conditions. What happens if two Elixir processes try to delete the same tensor at the same time? Do we have a simple and cheap solution in C/C++ where we would have a mutex/lock per tensor? Otherwise we could solve this on the Elixir side though and have this operation happen behind a pool of Elixir processes.

Reply:
good point. didn't realize NIF need to deal with multi threads. Is this the only function worry about threading?

Reply:
I think this is the only function performing mutations... or am I wrong and none of libtorch/at API is thread-safe?

Reply:
I found NIF has mutex APIs we can use.